### ISSUE TITLE
Unexpected metrics/labels missing in VictoriaMetrics

### ISSUE THREAD

**Issue by timur-ND** [VictoriaMetrics/VictoriaMetrics#6789]
### Describe the bug

This may be a long story, it will be difficult to describe, but I will try. I have 2 Victoria Metrics intances on two different servers in Docker, let's call them **Victoria1** and **Victoria2**. Data is sent there from two Prometheus and the remote_write configuration is the same. In general, all the parameters are the same. 
So, the problem is that **Victoria1** does not have some metrics, or rather, there are no metrics for some labels. For example, there is a metric `grpc_server_started_total`, it is in both Victorias, but with such labels there is no `grpc_server_started_total{hostname="myhost1.mycom ",grpc_service="etcdserverpb.Watch"} `on **Victoria1**. 
At the same time, on the second one it is (and on Prometheus too, the values ​​are the same `grpc_server_started_total{hostname="myhost1.mycom ",grpc_service="etcdserverpb.Watch"} 4878`)
at the same time, on Victoria1 this metric exists for the label `grpc_server_started_total{hostname="myhost2.mycom ",grpc_service="etcdserverpb.Watch"}` or even `grpc_server_started_total{hostname="myhost1.mycom"}`. 
In general, some metrics for specific labels are missing. In fact, in **Victoria1** data all these metrics were (they were approximately from April of this year until May, then disappeared for no reason). During this time I updated both Victorias several times (with deleting the cache https://docs.victoriametrics.com/url-examples/#internalresetrollupresultcache ), but nothing helped, new metrics did not appear in **Victoria1**. 
I actually checked everything I could from all sides, here is the full list:

- upgraded to latest (because found issues in Github regarding metrics missing, like https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5948 )
- checked Victoria1 logs, no errors
- startup flags the same on both Victorias
- tested changing some values for flags like **-dedup.minScrapeInterval** (maybe error with duplication, because both Prometheuses push data to both Victorias and I use global label for all metrics to avoid duplication - `prometheus: global`, but seems it's not the reason)
- checked Nginx error logs on Victoria1, no errors about upstream or any 5xx statuses (only during restart)
- Tested push to new instance. I created empty instance (same version and config flags) and push metrics to it - metrics appear
- Tested changing retention (I thought retention somehow removed metrics, but no effect) 
- Dropped cache (docker stop victoria1 && rm -rf cache/*), restart Victoria1, no effect
- Dropped cache via api `http://localhost:8428/internal/resetRollupResultCache `, no effect
- Exported metric from Prometheus and pushed to Victoria1 via import like `curl -v -X POST 'http://localhost:8428/api/v1/import/prometheus' -u '***' -d 'grpc_server_started_total{grpc_service="etcdserverpb.Watch",hostname="myhost1.mycom"} 4878'` , metric appeared for 10 minutes and was removed from Victoria1 due to unknown for me reason (no logs about that)
- Removed completely metric via `curl -u '***' -X POST 'http://localhost:8428/api/v1/admin/tsdb/delete_series' -d 'match[]=grpc_server_started_total{hostname="myhost1.mycom",grpc_service="etcdserverpb.Watch"}'`, no effect, no new metrics appeared for these labels
- Checked prometheus logs, no errors about both victorias, only resharding logs
- Reloaded prometheus
- Checked configs on prometheus side for pushing to Victorias, they are same fro both of them
- Changed parameters for victoria1 for pushing on prometheus side (like **max_shards** or **max_samples_per_send**), no effect
- Checked prometheus metrics about errors of sending data to Victoria, didn't find anything suspicious in **prometheus_remote_storage_*** metrics

Now I just don't know what else to do and how to return the metrics. In fact, this is not the only example, some metrics (again for some labels) are also missing, for example, the standard metrics from Node Exporter - `node_network_transmit_bytes_total `or `node_network_receive_bytes_total`.

I noticed the problem not so long ago, since I have Victoria1 as a backup and I do not want to delete data from it (about 10 months old). I can assume that it all started with version v1.99.0 (I updated it in March, metrics started to disappear, judging by the data, somewhere around the beginning of May).

### To Reproduce

I don’t know how to reproduce it, because the reason for the loss of metrics is unknown.

### Version

On both Victorias
```
./victoria-metrics-prod --version
victoria-metrics-20240801-125729-tags-v1.102.1-0-g996b623585
```
Prometheus - 2.50.1

### Logs

Prometheus logs contain only info about resharding:
tail -n 10000 /var/log/docker/prometheus.log | grep victoria1
```
2024-08-09T13:43:55.354544+00:00 prometheus1 docker/prometheus/3760cdc7f90e/prom/prometheus:v2.50.1[6122]: ts=2024-08-09T13:43:55.354Z caller=dedupe.go:112 component=remote level=info remote_name=90011f url=https://victoria1/api/v1/write msg="Remote storage resharding" from=4 to=6
2024-08-09T13:44:25.353904+00:00 prometheus1 docker/prometheus/3760cdc7f90e/prom/prometheus:v2.50.1[6122]: ts=2024-08-09T13:44:25.353Z caller=dedupe.go:112 component=remote level=info remote_name=90011f url=https://victoria1/api/v1/write msg="Remote storage resharding" from=6 to=4
2024-08-09T13:46:55.353246+00:00 prometheus1 docker/prometheus/3760cdc7f90e/prom/prometheus:v2.50.1[6122]: ts=2024-08-09T13:46:55.353Z caller=dedupe.go:112 component=remote level=info remote_name=90011f url=https://victoria1/api/v1/write msg="Remote storage resharding" from=4 to=6
2024-08-09T13:48:05.353303+00:00 prometheus1 docker/prometheus/3760cdc7f90e/prom/prometheus:v2.50.1[6122]: ts=2024-08-09T13:48:05.353Z caller=dedupe.go:112 component=remote level=info remote_name=90011f url=https://victoria1/api/v1/write msg="Remote storage resharding" from=6 to=4
```

Victoria1 logs contain only logs about cache and merges:
```
2024-08-09T13:41:13.914516+00:00 victoria1 docker/victoria-metrics/fb8602e58b4e/victoriametrics/victoria-metrics:v1.102.1[64720]: 2024-08-09T13:41:13.914Z#011info#011VictoriaMetrics/lib/storage/partition.go:1465#011merged (15 parts, 231176199 rows, 176143081 blocks, 2411252145 bytes) into (1 part, 113561638 rows, 19503601 blocks, 431578105 bytes) in 70.846 seconds at 3263098 rows/sec to "/victoria-metrics-data/data/small/2024_08/17E9C1D30055DC99"
2024-08-09T13:41:18.168612+00:00 victoria1 docker/victoria-metrics/fb8602e58b4e/victoriametrics/victoria-metrics:v1.102.1[64720]: 2024-08-09T13:41:18.168Z#011warn#011VictoriaMetrics/app/vmselect/promql/rollup_result_cache.go:77#011resetting rollup result cache because the metric container_network_transmit_errors_total{job=... #omited} (Timestamp=1723210563931, Value=0.000000) has a timestamp older than -search.cacheTimestampOffset=5m0s by 13.069s
2024-08-09T13:41:18.168969+00:00 victoria1 docker/victoria-metrics/fb8602e58b4e/victoriametrics/victoria-metrics:v1.102.1[64720]: 2024-08-09T13:41:18.168Z#011info#011VictoriaMetrics/app/vmselect/promql/rollup_result_cache.go:216#011rollupResult cache has been cleared
2024-08-09T13:41:45.697487+00:00 victoria1 docker/victoria-metrics/fb8602e58b4e/victoriametrics/victoria-metrics:v1.102.1[64720]: 2024-08-09T13:41:45.697Z#011info#011VictoriaMetrics/lib/storage/partition.go:1465#011merged (10 parts, 84359258 rows, 83947992 blocks, 1055850569 bytes) into (1 part, 35761865 rows, 19038960 blocks, 307431145 bytes) in 36.028 seconds at 2341493 rows/sec to "/victoria-metrics-data/data/small/2024_08/17E9C1D30055DCA4"
```
Almost the same on another Victoria. If I skip all these logs to find something, then nothing:
`cat /var/log/docker/victoria-metrics.log  | grep -Ev "rollupResult cache has been cleared|search.cacheTimestampOffset|011merged"`

Nginx error log only about buffering (on both Victorias), so I don't think it's somehow related:
`2024/08/09 00:03:41 [warn] 2586365#2586365: *94781939 a client request body is buffered to a temporary file /dev/shm/client_body_temp_path/11/71/0093277111, client: 10.96.1.79, server: victoria1, request: "POST /api/v1/write HTTP/1.1", host: "victoria1"`

### Screenshots

<img width="1422" alt="Screenshot 2024-08-09 at 16 21 02" src="https://github.com/user-attachments/assets/26820e95-5060-46e0-b386-517c7baca210">
<img width="1427" alt="Screenshot 2024-08-09 at 16 21 29" src="https://github.com/user-attachments/assets/3b71a28e-26b9-46fb-87e1-6f863e93c3ae">
<img width="1429" alt="Screenshot 2024-08-09 at 17 06 52" src="https://github.com/user-attachments/assets/fc0e2190-db9e-4ebf-a220-02cf63516ec3">

Screenshots from both Victorias and Prometheus, first - from Victoria1, second - from Victoria2, third - from Prometheus

### Used command-line flags

Docker command-line flags (only retention is different, on Victoria2 is 365d, but no such amount of metrics for now):
```
        "Args": [
            "-httpListenAddr",
            "127.0.0.1:8428",
            "-retentionPeriod",
            "1095d",
            "-httpAuth.username",
            "***",
            "-httpAuth.password",
            "***",
            "-search.maxUniqueTimeseries=10000000",
            "-search.minStalenessInterval=5m",
            "-dedup.minScrapeInterval=60s",
            "-maxLabelValueLen=30000",
            "-search.latencyOffset=1m",
            "-maxLabelsPerTimeseries=50"
        ],
```

### Additional information

Prometheus config for push (maybe problem somewhere here):
```
remote_write:
  - url: https://victoria1/api/v1/write
    basic_auth:
      username: ***
      password: ***
    queue_config:
      capacity: 150000
      max_samples_per_send: 50000
      max_shards: 75
    metadata_config:
      send_interval: 45s
      max_samples_per_send: 50000
  - url: https://victoria2/api/v1/write
    basic_auth:
      username: ***
      password: ***
    queue_config:
      capacity: 150000
      max_samples_per_send: 50000
      max_shards: 75
    metadata_config:
      send_interval: 45s
      max_samples_per_send: 50000
```

**Comment by Amper** [comment]
Hello @timur-ND.
Unfortunately v1.99.0 and v1.100.0 contain serious bugs that could cause described effect  (missing metrics). See notes for these versions:

> **This release contains [the issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5959), which can prevent from storing data for new time series under high rate of search queries. Please rollback to [v1.98.0](https://docs.victoriametrics.com/changelog/#v1980) or upgrade to [v1.101.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.101.0).**

Also you can see details in related issues:

- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5948
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5959

**Comment by timur-ND** [comment]
Yes, I probably read these issues, and I upgraded to latest) As you see, my current version is v1.102.1
Regarding this https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5948 ,after read, I had v1.99.0, I upgraded to v1.100.0 (with dropping cache). Then upgraded to v1.100.1, no effect. Then to current latest - v1.102.1

**Comment by timur-ND** [comment]
regarding https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5959 , I see increasing of `vm_missing_metric_names_for_metric_id_total` as it was described there. Time of metric increasing correlates with my metrics first missing( which I described above)
<img width="1413" alt="Screenshot 2024-08-09 at 18 28 57" src="https://github.com/user-attachments/assets/95933fa0-2611-44df-a569-245e152fd50b">

**Comment by hagen1778** [comment]
@timur-ND do you observe these spikes on updated version? Or these are historical data?

**Comment by timur-ND** [comment]
@hagen1778 regarding `vm_missing_metric_names_for_metric_id_total` no, I don't see increasing now. In the screenshot above, spikes were several months ago. At that time I changed retention so probably some old metrics were deleted due old retention period. I have a data from 1st October 2023, so in the picture spikes started in the beginning of May 2024, and at that time retention flag was `-retentionPeriod 180d`. I changed to 365d and no more spikes, so probably it's just related to retention. I found in the linked issues a mention about that metric, that's why I also checked it.

little increasing of `vm_missing_metric_names_for_metric_id_total` 2 weeks ago - I tested importing problematic metrics to `api/v1/import/prometheus` and Victoria deleted that data (I mentioned it in the issue description)

**Comment by hagen1778** [comment]
Thanks!
> Exported metric from Prometheus and pushed to Victoria1 via import like curl -v -X POST 'http://localhost:8428/api/v1/import/prometheus' -u '***' -d 'grpc_server_started_total{grpc_service="etcdserverpb.Watch",hostname="myhost1.mycom"} 4878' , metric appeared for 10 minutes and was removed from Victoria1 due to unknown for me reason (no logs about that)

Did you make this experiment on the updated version of VM (>1.99)? Is series `grpc_server_started_total{grpc_service="etcdserverpb.Watch",hostname="myhost1.mycom"}` currently active and is ingested/updated for Victoria2?

**Comment by timur-ND** [comment]
Yes, I've just made experiment again on the latest updated version, import is successful, but metric has been removed after 10 minutes as before:
<img width="1421" alt="Screenshot 2024-08-16 at 11 43 08" src="https://github.com/user-attachments/assets/2371589f-9b22-47c7-ae99-20c66f178e49">
Nothing is logs about that: `cat /var/log/docker/victoria-metrics.log| grep grpc`

Here, on Victoria2, metric exists with the same labels, active and receive new values:
<img width="1433" alt="Screenshot 2024-08-16 at 11 45 41" src="https://github.com/user-attachments/assets/74e70be0-7048-4c28-9997-aea0b57d0763">

**Comment by hagen1778** [comment]
> but metric has been removed after 10 minutes as before, I mean no new series for this metric

Please correct me if I'm wrong, but this doesn't look like metric was removed. It is still there, screenshot shows it is queryable. And it is likely contains only samples you manually imported. Which makes me think that for some reason new samples doesn't arrive or are ignored by VM...

Can we do a little experiment by making Prometheus to remote-write to vmagent? And vmagent will replicate data to Victoria1 and Victoria2

**Comment by timur-ND** [comment]
> Please correct me if I'm wrong, but this doesn't look like metric was removed. It is still there, screenshot shows it is queryable. And it is likely contains only samples you manually imported. Which makes me think that for some reason new samples doesn't arrive or are ignored by VM.

Yes, you are right. Metric exists and looks like VM doesn't accept new values from remote-write from Prometheus

About vmagent, I have never tried it, so I need sometime to check it and configure. Can I use vmagent to push the same metrics to VMs simultaneously with Prometheus? Or I need to stop remote_write and only then use vmagent?

I will come back later with the results

**Comment by timur-ND** [comment]
Ok, after vmagent tests, I see the metric has appeared.
<img width="1422" alt="Screenshot 2024-08-19 at 12 47 29" src="https://github.com/user-attachments/assets/46643b46-067b-4f26-a971-085b358ce98e">

Now I'm completely confused why I don't see the same using Prometheus.

`vmagent-prod --version
vmagent-20240801-125405-tags-v1.102.1-0-g996b623585`

vmagent startup flags:
`vmagent-prod -promscrape.config vmagent.yaml  -remoteWrite.url https://***:***@victoria1/api/v1/write -promscrape.config.strictParse=false`

vmagent.yaml:
```
scrape_configs:
- job_name: "etcd"
  static_configs:
  - targets:
    - "127.0.0.1:2381"
    labels:
      __app_name: etcd
      __metrics_path__: "/metrics"
      hostname: "dra4-c2-169-239-72.fe.gc.onl"

remote_write:
  - url: https://victoria1/api/v1/write
    basic_auth:
      username: ***
      password: ***
 # reducing params because a few amount of metrics
    queue_config:
      capacity: 1500
      max_samples_per_send: 500
      max_shards: 10
    metadata_config:
      send_interval: 45s
      max_samples_per_send: 500
```

Nginx log:
```
***** - vm [19/Aug/2024:09:40:12 +0000] "POST /api/v1/write HTTP/1.1" 204 0 "-" "vmagent"
***** - vm [19/Aug/2024:09:41:13 +0000] "POST /api/v1/write HTTP/1.1" 204 0 "-" "vmagent"
***** - vm [19/Aug/2024:09:42:12 +0000] "POST /api/v1/write HTTP/1.1" 204 0 "-" "vmagent"
```

And as I mentioned, problem can be not only in that metric, faced the same behaviour with another metrics

**Comment by hagen1778** [comment]
> Now I'm completely confused why I don't see the same using Prometheus.

Can you tell what Prometheus version you used, so I could take a look at the code?

> Now I'm completely confused why I don't see the same using Prometheus.

It is unclear to me as well. Either Prometheus doesn't replicate data correctly, or maybe it adds some extra labels or modifications to the sent series, so it gets dropped or just looks differently. Not sure, have to read Prom code.

As a workaround I'd recommend replacing Prometheus with vmagent for now.

**Comment by timur-ND** [comment]
> Can you tell what Prometheus version you used, so I could take a look at the code?

Currently 2.50.1, but I want to update it to latest in the near future

**Comment by timur-ND** [comment]
Also interesting moment I described above:

> `Tested push to new instance. I created empty instance (same version and config flags) and push metrics to it - metrics appear`

Here it seems the issue on VM side, because metrics appeared in the new instance. But after vmagent tests:

> Ok, after vmagent tests, I see the metric has appeared.

It seems problem on prometheus side. I am confused where the problem...
Also I dropped entire prometheus tsdb (not so much data was here, since data exist in VM), in the hope that it will help, but the result is the same. I will update prometheus on next week, but now I'm not sure if this will help

**Comment by hagen1778** [comment]
> Tested push to new instance. I created empty instance (same version and config flags) and push metrics to it - metrics appear

@timur-ND how exactly did you push data to this new instance? Was it a 3rd remote write url? Or you substituted one of remote-write urls you already had with new address?

Also could you please share what's the ingestion rate for VM instances?

**Comment by timur-ND** [comment]
@hagen1778 I added new endpoint to remote_write section with same config on one Prometheus server (remember I have two Prometheus servers with the same configs). 

```
  - url: https://victoria-test/api/v1/write
    basic_auth:
      username: ***
      password: ***
    queue_config:
      capacity: 150000
      max_samples_per_send: 50000
      max_shards: 75
    metadata_config:
      send_interval: 45s
      max_samples_per_send: 50000
```
I also tested case if there is some kind of metrics duplication - shut down one prometheus - no effect, no metrics in VM1

About ingestion rate, here for last 30 days:

Blue color - VM2 (good)
Red color - VM1 (not good)

`rate(vm_rows_inserted_total{type="promremotewrite"}[1m])` rate is about 1.5M, but for VM2 there are some spikes about 1.8M and higher.

<img width="1432" alt="Screenshot 2024-08-23 at 16 09 13" src="https://github.com/user-attachments/assets/da7b414a-e1b5-4c45-a028-5dddff8cb66a">

`rate(vm_http_requests_total{path="/api/v1/write"}[1m])` rate is about 30 for both VMS

<img width="1423" alt="Screenshot 2024-08-23 at 16 16 41" src="https://github.com/user-attachments/assets/00e4606d-cce1-4cd8-abc6-ec8968fa8077">

**Comment by f41gh7** [comment]
Hello, could please check prometheus remote write delay by checking the following metrics at prometheus side:
`prometheus_remote_storage_highest_timestamp_in_seconds` and `prometheus_remote_storage_queue_highest_sent_timestamp_seconds`. 

 Log lines with `rollupResult cache has been cleared` messages look suspicious to me. Usually it indicates about historical data back-filling or ingestion delay at remote client.

**Comment by timur-ND** [comment]
If I understood right, you asked about difference between these metrics. Here my query for VM1:

`sum by (instance)(prometheus_remote_storage_highest_timestamp_in_seconds{instance=~".*-prometheus-.*"}) - sum by (instance)( prometheus_remote_storage_queue_highest_sent_timestamp_seconds{instance=~".*-prometheus-.*", url="https://victoria1/api/v1/write"})`
<img width="1426" alt="Screenshot 2024-08-26 at 11 17 09" src="https://github.com/user-attachments/assets/7e79f8d9-898c-4a32-b392-6d5a2f50bc8e">

And for VM2:
<img width="1425" alt="Screenshot 2024-08-26 at 11 19 29" src="https://github.com/user-attachments/assets/ad6a43b9-869c-46dc-bafa-9838b2068dad">

Results:
VM1 - max difference 10 seconds
VM2 - max difference 43 seconds

For VM2 looks worse than for VM1, but VM2 is good with metrics) I think, that big difference for VM2 somehow related to the load, CPU load is higher and this server has less cpu cores, but maybe I am wrong. Anyway, for VM1 situation with such metrics are better( correct me please if I am wrong).

Regarding `rollupResult cache has been cleared` - I also see this log for both VMs, and all these logs related to k8s metrics, like this:

```
2024-08-26T08:36:06.188Z	warn	VictoriaMetrics/app/vmselect/promql/rollup_result_cache.go:77	resetting rollup result cache because the metric container_cpu_load_average_10s{namespace="colocation-backend-preprod",job="kubernetes-nodes-cadvisor",instance="***",pod="colocation-sync-megaport-disabled-vxc-28744350-lg2fd",name="e8db4688034b74b06fbbd2e6db2694e1b9008677166cb3e3d4fc06ad115292e4",beta_kubernetes_io_arch="amd64",beta_kubernetes_io_os="linux",id="/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod622bbb84_101b_420f_a181_2d62c971306a.slice/cri-containerd-e8db4688034b74b06fbbd2e6db2694e1b9008677166cb3e3d4fc06ad115292e4.scope",image="k8s.gcr.io/pause:3.6",k8s_cluster="prod",kubernetes_io_arch="amd64",kubernetes_io_hostname="***",kubernetes_io_os="linux"} (Timestamp=1724661064321, Value=0.000000) has a timestamp older than -search.cacheTimestampOffset=5m0s by 1.679s
2024-08-26T08:36:06.188Z	info	VictoriaMetrics/app/vmselect/promql/rollup_result_cache.go:216	rollupResult cache has been cleared
```

I don't know, maybe it's just because k8s metrics are not collected very often. Should I increase `-search.cacheTimestampOffset=10m0s` in that case? (maybe this is not related to the current problem)

**Comment by timur-ND** [comment]
Hello. I updated Prometheus to v2.54.1 and seems no effect. Metrics, which I provided in the issue description, appeared for unknown reason in the end of August. Don't know why, but I didn't do anything at that time, no restarts from Prometheus or VM side. I thought issue was fixed somehow, but found another metrics, which exist in VM2, but don't exist in VM1. So upgrade didn't help me, as I expected. Not sure it's useful information, but just let you know that seems not related to metrics sender version (Prometheus or Prom-Agent)

**Comment by timur-ND** [comment]
Seems the issue was due to data corruption long time ago. VM1 and VM2 were created long time ago, and VM1 was moved to another servers several times (to not lost metrics). Seems that the data was moved incorrectly (like rsync while Victoria was running). Before I noticed metrics difference, I also moved data from old VM1 server to the new, so corrupted data also was moved. 
So, I synced data from VM2 (good instance) to VM1 (bad instance) using rsync (while VM2 was stopped). Now I don't see metrics missing as provided in the issue description and new metrics continue to be saved. If it's true, strange that no logs about data corruption or something. No exact evidences that data was corrupted, just an assumption, but looks like the true, at least from my side. 
if it was, maybe there are some checks for data consistency from VM side? And put it explicitly to the logs, if any corruption appears.
If data synchronization was completed incorrectly, I saw that:
`panic: FATAL: part "/victoria-metrics-data/indexdb/17647E6429E8EF17/17E8BE40FF287598" is listed in "/victoria-metrics-data/indexdb/17647E6429E8EF17/parts.json", but is missing on disk; ensure "/victoria-metrics-data/indexdb/17647E6429E8EF17/parts.json" contents is not corrupted; remove "/victoria-metrics-data/indexdb/17647E6429E8EF17/parts.json" to rebuild its' content from the list of existing parts' content from the list of existing parts`
but in this case VM won't start. But in my case VM worked good all the time. Either data corruption wasn't found for long time or something else...
I will continue monitor metrics in both VMs and let you know later
